src/tile.rs

```diff
@@ -2,26 +2,187 @@ use bytes::Bytes;
+    /// # endianness: Endianness::LittleEndian,
+    ///   image_width: 15,
+    ///   image_height: 15,
+    ///   chunk_width: 8,
+    ///   chunk_height: 8,
+    /// # bits_per_sample: &[32],
+    /// # samples_per_pixel: 1,
+    /// # sample_format: &[SampleFormat::IEEEFP],
+    /// # planar_configuration: PlanarConfiguration::Chunky,
+    /// };
+    ///
+    /// assert_eq!(info.chunk_height_pixels(0).unwrap(), (8));
+    /// assert_eq!(info.chunk_height_pixels(1).unwrap(), (7));
+    /// info.chunk_height_pixels(2).unwrap_err();
+    /// ```
+    pub fn chunk_height_pixels(&self, y: u32) -> AsyncTiffResult<u32> {
+        if y >= self.chunks_down() {
+            Err(crate::error::AsyncTiffError::TileIndexError(
+                y,
+                self.chunks_down(),
+            ))
+        } else if y == self.chunks_down() - 1 {
+            // last chunk
+            Ok(self.image_height - self.chunk_height * y)
+        } else {
+            Ok(self.chunk_height)
+        }
+    }
+
+    /// get the output row stride in bytes, taking padding into account
+    pub fn output_row_stride(&self, x: u32) -> AsyncTiffResult<usize> {
+        Ok((self.chunk_width_pixels(x)? as usize).saturating_mul(self.bits_per_pixel()) / 8)
+    }
+
+    /// The total number of bits per pixel, taking into account possible different sample sizes
+    ///
+    /// Technically bits_per_sample.len() should be *equal* to samples, but libtiff also allows
+    /// it to be a single value that applies to all samples.
+    ///
+    /// Libtiff and image-tiff do not support mixed bits per sample, but we give the possibility
+    /// unless you also have PlanarConfiguration::Planar, at which point the first is taken
+    pub fn bits_per_pixel(&self) -> usize {
+        match self.planar_configuration {
+            PlanarConfiguration::Chunky => {
+                if self.bits_per_sample.len() == 1 {
+                    self.samples_per_pixel as usize * self.bits_per_sample[0] as usize
+                } else {
+                    assert_eq!(self.samples_per_pixel as usize, self.bits_per_sample.len());
+                    self.bits_per_sample.iter().map(|v| *v as usize).product()
+                }
+            }
+            PlanarConfiguration::Planar => self.bits_per_sample[0] as usize,
+        }
+    }
+
+    /// The number of chunks in the horizontal (x) direction
+    pub fn chunks_across(&self) -> u32 {
+        self.image_width.div_ceil(self.chunk_width)
+    }
+
+    /// The number of chunks in the vertical (y) direction
+    pub fn chunks_down(&self) -> u32 {
+        self.image_height.div_ceil(self.chunk_height)
+    }
+}
+
 /// A TIFF Tile response.
 ///
 /// This contains the required information to decode the tile. Decoding is separated from fetching
 /// so that sync and async operations can be separated and non-blocking.
 ///
 /// This is returned by `fetch_tile`.
 #[derive(Debug)]
-pub struct Tile {
+pub struct Tile<'a> {
```

Comment:
Given that we have Python bindings, and we expose `Tile` to Python, we can't have a lifetime specifier here. It's not possible to have a `#[pyclass]` with internal lifetimes. 

Instead, we should remove any borrowed data from `Tile` and `PredictorInfo` and store `Arc<PredictorInfo>` here.